### DISCUSSION TITLE
How to use offset pagination with Prisma

### DISCUSSION THREAD

**Discussion by adherb** [prisma/prisma#11405] [Q&A]
I am trying to upgrade my pagination to use Prisma offset pagination. I am currently storing all posts from an api call in an array and using this to paginate. What I would like to do is use Prisma to pass 'take' and 'skip' dynamically.

Currently I have the following:
```jsx
//index.js
import React, { useState, useEffect } from 'react';
import Posts from './components/Posts';
import Pagination from './components/Pagination';

const App = () => {
  const [posts, setPosts] = useState([]);
  const [loading, setLoading] = useState(false);
  const [currentPage, setCurrentPage] = useState(1);
  const [postsPerPage] = useState(10);

  useEffect(() => {
    const fetchPosts = async () => {
      setLoading(true);
      const res = await fetch`${process.env.NEXT_PUBLIC_SERVER_API}/posts/get`,
      setPosts(res.data);
      setLoading(false);
    };

    fetchPosts();
  }, []);

  // Get current posts
  const indexOfLastPost = currentPage * postsPerPage;
  const indexOfFirstPost = indexOfLastPost - postsPerPage;
  const currentPosts = posts.slice(indexOfFirstPost, indexOfLastPost);

  // Change page
  const paginate = pageNumber => setCurrentPage(pageNumber);

  return (
    <div className='container mt-5'>
      <h1 className='text-primary mb-3'>My Blog</h1>
      <Posts posts={currentPosts} loading={loading} />
      <Pagination
        postsPerPage={postsPerPage}
        totalPosts={posts.length}
        paginate={paginate}
      />
    </div>
  );
};

export default App;
```

```jsx
//Pagination Component
import React from 'react';

const Pagination = ({ postsPerPage, totalPosts, paginate }) => {
  const pageNumbers = [];

  for (let i = 1; i <= Math.ceil(totalPosts / postsPerPage); i++) {
    pageNumbers.push(i);
  }

  return (
    <nav>
      <ul className='pagination'>
        {pageNumbers.map(number => (
          <li key={number} className='page-item'>
            <a onClick={() => paginate(number)} href='!#' className='page-link'>
              {number}
            </a>
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default Pagination;
```

What I would like to do is use [Prisma's offset pagination][1], dynamically passing take and skip.

So the use effect would have skip and take params added like this:
```jsx
useEffect(() => {
    const fetchPosts = async () => {
      setLoading(true);
      const res = await fetch`${process.env.NEXT_PUBLIC_SERVER_API}/posts/get take=15&skip=15`,
      setPosts(res.data);
      setLoading(false);
    };

    fetchPosts();
  }, []);
```

Here is my Prisma query:
```ts
const posts = await prisma.post.findMany({
      take: parseInt(take),
      skip: parseInt(skip),
      select: {
        id: true,
        title: true,
      },
      orderBy: {
        title: "asc",
      },
    });
    res.json(posts)
```

How can I dynamically pass and track pagination using take and skip on the frontend?

  [1]: https://www.prisma.io/docs/concepts/components/prisma-client/pagination

**Comment by austincrim** [comment]
Hey @adherb 👋 ,

Just to clarify, are you asking how to setup your React frontend to send the correct parameters to your Prisma query? The Prisma query you posted looks good to me, for what it's worth.

[From our docs](https://www.prisma.io/docs/concepts/components/prisma-client/pagination):
> To implement pages of results, you would just skip the number of pages multiplied by the number of results you show per page.

If you already have a set number of results per page, you then need to send the current page number to the backend. With that info, you multiply your page number by your amount of results per page, and that's your `skip` param.

Something like this:
```ts
await prisma.post.findMany({
  skip: (resultsPerPage * currentPageNumber),
  take: resultsPerPage,
  ...
})
```

You could pass this info in query parameters, like you showed in your snippet.

Let me know if I can clarify anything further!

**Reply by vladi-stevanovic** [reply]
👋 Hello @adherb I'm going through pending GitHub Discussions to check if you need any help. Is this issue still open?
Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! :raised_hands:

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
